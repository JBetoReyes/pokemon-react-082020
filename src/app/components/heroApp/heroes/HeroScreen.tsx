/* eslint-disable @typescript-eslint/naming-convention */
import React, { useMemo } from 'react';
import { useParams, Redirect, RouteComponentProps } from 'react-router-dom';
import { AppClickEvent } from '@typings/htmlEvents';
import getHeroesById from '../services/getHeroesById';
import { IHero } from '../data/heroes';

type Props = Pick<RouteComponentProps, 'history'>;

const HeroScreen = ({ history }: Props): JSX.Element => {
  const { heroId } = useParams<{ heroId: string }>();
  const hero = useMemo<IHero | undefined>(() => getHeroesById(heroId), [
    heroId,
  ]);
  if (!hero) {
    return <Redirect to="/" />;
  }
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const handleClick = (e: AppClickEvent) => {
    if (history.length <= 2) {
      history.push('/');
    }
    history.goBack();
  };
  const {
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,
  } = hero as IHero;
  return (
    <div className="row m-5">
      <div className="col-3">
        <img
          src={`../heroes/${heroId}.jpg`}
          alt={superhero}
          className="img-thumbnail"
        />
      </div>
      <div className="col-8">
        <h3>{superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            {' '}
            <b> Alter ego: </b>
            {alter_ego}
          </li>
          <li className="list-group-item">
            {' '}
            <b> Publisher: </b>
            {publisher}
          </li>
          <li className="list-group-item">
            {' '}
            <b> First appearance: </b>
            {first_appearance}
          </li>
        </ul>

        <h5> Characters </h5>
        <p>{characters}</p>

        <button
          type="button"
          className="btn btn-outline-info"
          onClick={handleClick}
        >
          Return
        </button>
      </div>
    </div>
  );
};

export default HeroScreen;
