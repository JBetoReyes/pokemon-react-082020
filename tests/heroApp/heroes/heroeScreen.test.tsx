/* eslint-disable @typescript-eslint/no-explicit-any */
import HeroScreen from '@components/heroApp/heroes/HeroScreen';
import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';

describe('heroe screen tests', () => {
  test('Should match the snapshot', () => {
    const mockHistory = {
      push: jest.fn(),
      goBack: jest.fn(),
      length: 3,
    };
    const wrapper = mount(
      <MemoryRouter initialEntries={['/heroes']}>
        <HeroScreen history={mockHistory as any} />
      </MemoryRouter>
    );
    expect(wrapper.find('Redirect')).toMatchSnapshot();
    expect(wrapper.find('Redirect').exists()).toBe(true);
  });
  test('Should render the heroe data', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Route path="/hero/:heroId" component={HeroScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find('.row').exists()).toBe(true);
  });
  test('Should push the root path if the history length is less or equal two', () => {
    const mockHistory = {
      push: jest.fn(),
      goBack: jest.fn(),
      length: 2,
    };
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Route
          path="/hero/:heroId"
          component={() => <HeroScreen history={mockHistory as any} />}
        />
      </MemoryRouter>
    );
    wrapper.find('button').simulate('click');
    expect(mockHistory.push).toBeCalledWith('/');
  });
  test('Should go back to the previous screen', () => {
    const mockHistory = {
      push: jest.fn(),
      goBack: jest.fn(),
      length: 3,
    };
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider']}>
        <Route
          path="/hero/:heroId"
          component={() => <HeroScreen history={mockHistory as any} />}
        />
      </MemoryRouter>
    );
    wrapper.find('button').prop('onClick')({} as any);
    expect(mockHistory.goBack).toHaveBeenCalled();
  });
  test('Should not render the coomponent with an invalid id', () => {
    const mockHistory = {
      push: jest.fn(),
      goBack: jest.fn(),
      length: 3,
    };
    const wrapper = mount(
      <MemoryRouter initialEntries={['/hero/marvel-spider1']}>
        <Route
          path="/hero/:heroId"
          component={() => <HeroScreen history={mockHistory as any} />}
        />
      </MemoryRouter>
    );
    expect(wrapper.find('.row').exists()).toBe(false);
  });
});
