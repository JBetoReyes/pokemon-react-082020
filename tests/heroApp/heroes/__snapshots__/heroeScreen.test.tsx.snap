// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`heroe screen tests Should match the snapshot 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/heroes",
    ]
  }
>
  <Router
    history={
      Object {
        "action": "REPLACE",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "eki9g7",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "eki9g7",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <HeroScreen
      history={
        Object {
          "goBack": [MockFunction],
          "length": 3,
          "push": [MockFunction],
        }
      }
    >
      <Redirect
        to="/"
      >
        <Lifecycle
          onMount={[Function]}
          onUpdate={[Function]}
          to="/"
        />
      </Redirect>
    </HeroScreen>
  </Router>
</MemoryRouter>
`;

exports[`heroe screen tests Should render the heroe data 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/hero/marvel-spider",
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "uawkkj",
            "pathname": "/hero/marvel-spider",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "uawkkj",
          "pathname": "/hero/marvel-spider",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Route
      component={[Function]}
      path="/hero/:heroId"
    >
      <HeroScreen
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "uawkkj",
                "pathname": "/hero/marvel-spider",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "uawkkj",
              "pathname": "/hero/marvel-spider",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "key": "uawkkj",
            "pathname": "/hero/marvel-spider",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {
              "heroId": "marvel-spider",
            },
            "path": "/hero/:heroId",
            "url": "/hero/marvel-spider",
          }
        }
      >
        <div
          className="row m-5"
        >
          <div
            className="col-3"
          >
            <img
              alt="Spider Man"
              className="img-thumbnail"
              src="../heroes/marvel-spider.jpg"
            />
          </div>
          <div
            className="col-8"
          >
            <h3>
              Spider Man
            </h3>
            <ul
              className="list-group list-group-flush"
            >
              <li
                className="list-group-item"
              >
                 
                <b>
                   Alter ego: 
                </b>
                Peter Parker
              </li>
              <li
                className="list-group-item"
              >
                 
                <b>
                   Publisher: 
                </b>
                Marvel Comics
              </li>
              <li
                className="list-group-item"
              >
                 
                <b>
                   First appearance: 
                </b>
                Amazing Fantasy #15
              </li>
            </ul>
            <h5>
               Characters 
            </h5>
            <p>
              Peter Parker
            </p>
            <button
              className="btn btn-outline-info"
              onClick={[Function]}
              type="button"
            >
              Return
            </button>
          </div>
        </div>
      </HeroScreen>
    </Route>
  </Router>
</MemoryRouter>
`;
